### PR TITLE
Support for AVX512

### PR DESCRIPTION
**Pull request by abadams**
Assumes the cannonlake / skylake server feature set. Relies on Intel's SDE emulator for testing.

I only found one llvm bug doing this (https://llvm.org/bugs/show_bug.cgi?id=30947) so I think llvm is sufficiently mature for this now.

See issue #1608 for the feature request

### CODE REVIEW

**Comment by jrk** [comment]
Curious: what SKL/CNL-specific features does this assume? (They do have a
bunch of AVX512 ops not in KNL/Phi x200.)

File: src/runtime/HalideRuntime.h [hunk 1]
```diff
@@ -723,30 +723,28 @@ typedef enum halide_target_feature_t {
     halide_target_feature_sse41 = 4,  ///< Use SSE 4.1 and earlier instructions. Only relevant on x86.
     halide_target_feature_avx = 5,  ///< Use AVX 1 instructions. Only relevant on x86.
     halide_target_feature_avx2 = 6,  ///< Use AVX 2 instructions. Only relevant on x86.
-    halide_target_feature_fma = 7,  ///< Enable x86 FMA instruction
-    halide_target_feature_fma4 = 8,  ///< Enable x86 (AMD) FMA4 instruction set
-    halide_target_feature_f16c = 9,  ///< Enable x86 16-bit float support
+    halide_target_feature_avx512 = 7,  ///< Use AVX 512 instructions (the subset supported by Cannonlake). Only relevant on x86.
```
Comment (steven-johnson) [inline src/runtime/HalideRuntime.h]: Do we really want to renumber these? I thought we had decided to avoid doing so when possible.

Comment (abadams) [inline src/runtime/HalideRuntime.h]: It seemed like a good idea to put avx512 next to avx2. I could also move it to the end.

**Comment by abadams** [comment]
The feature set assumed here is: +avx512f,+avx512cd,+avx512bw,+avx512dq,+avx512vl,+avx512er,+avx512ifma,+avx512pf,+avx512vbmi

We can get more granular pretty easily with another target flag. Is it worth breaking out the KNL subset?

**Comment by jrk** [comment]
I was just assuming that KNL was going to be a target of interest to some people.

- Everyone has F + CD + ER
- KNL has only those + PF, which SKL and CNL *don't* have
- SKL-EP adds VL + BW + DQ
- CNL = [SKL-EP] + IFMA + VBMI

So no CPUs have the full set listed, if I'm understanding correctly (SKL & CNL don't have PF), and 2016-2017 chips (KNL & SKL) each lack at least several others.

([ref](https://en.wikipedia.org/wiki/AVX-512#CPUs_with_AVX-512))

**Comment by shoaibkamil** [comment]
In fact, I'm not sure that table on the Wikipedia page is correct; the linked presentation from Intel (from 2014) says ER and PF *both* will be Xeon Phi exclusive (i.e. SKL-EP won't have them).  See slide 3 [here](https://gcc.gnu.org/wiki/cauldron2014?action=AttachFile&do=get&target=Cauldron14_AVX-512_Vector_ISA_Kirill_Yukhin_20140711.pdf).

Regardless, @jrk's meta-point here stands: there's no processor with the set of features assumed here and they need to be broken out somehow.

**Comment by abadams** [comment]
Breaking it down by architecture seems most useful. How about the following four feature flags then:

AVX512 => F + CD
AVX512_KNL => AVX512 + PF + ER
AVX512_SKL => AVX512 + BL  + BW + DQ
AVX512_CNL => AVX512_SKL + IFMA + VBMI

**Comment by abadams** [comment]
I pushed a version with those four target features.

**Comment by shoaibkamil** [comment]
Four target features makes sense to me.

**Comment by zvookin** [comment]
I really do not like this approach to feature flags. The cognitive load here is huge. One has to know the expansion of the three letter acronym of the codename for the architecture, one has to know the mapping of that codename to actual chips which is not always straightforward. The list will grow over time. Now part of this is that Intel's design here is just horrible computer architecture practice and there's really nothing to be done for the complexity because there really are just too damn many variants of the ISA.

That said, some things we can do:

1) Surface the feature flags directly so if one knows the ISA subsets one wants, one can get them directly.
2) Introduce something to the naming convention such that one can tell AVX512_SKL is a subset of AVX512_CNL and that AVX_KNL is not and the rough chronology is AVX512_KNL, AVX512_SKL, AVX512_CNL.
3) The most important thing here I'd expect is whether one can vectorize integer ops or not, which I think is BW. It may be difficult to put this into the name, but we can definitely add more comments here.

I think concretely, my suggestion on the PR would be to surface each ISA option as a feature flag, keep the architecture ones as well, but maybe spelling out the names and adding a year of introduction. And really add a lot of text to the comments. (Spelling out the names is suggested as a one character diff between the options is problematic. If we add the year or a counter, that may not be necessary.)

This feedback is driven by having spent many hours of my life trying to find just the right compiler flags to target a given version of a chip. Usually I know stuff from the architecture manual pretty definitively, but things like codenames of specific parts require searching. It gets horridly worse five years down the road from when the chip actually shipped.

**Comment by abadams** [comment]
Surfacing each option as a feature flag would be a much larger list with
lots of redundancies in it. E.g. VBMI always goes with IFMA. As a dev, I
don't think this distinction is helpful. I think you know which generation
of chip you want to emit code for, not what Intel's codenames for feature
subsets are. I don't think there's a useful year to give either, given the
slow roll out of different feature sets at different times.

The names llvm uses for these cpus are "knl", "skylake-avx512", and
"cannonlake". Confusingly, there's also skylake *without* avx512. Perhaps
we should go with "avx512_knl", "avx512_skylake", and "avx512_cannonlake"
as feature names?

We can certainly include a longer description years and such in the
comments.

**Comment by shoaibkamil** [comment]
> The names llvm uses for these cpus are "knl", "skylake-avx512", and
> "cannonlake". Confusingly, there's also skylake *without* avx512.

Just an aside here-- the reason for this is desktop/mobile Skylake doesn't have AVX-512, only the Xeons versions (Skylake-EP/Skylake-EX) do.

**Comment by zvookin** [comment]
Fixing the one letter diff via spelling out the names as llvm does is certainly an improvement. Being consistent with llvm is likely also a win.

If we were to do feature flags, I have no issue with only providing one feature for combinations that only ever appear as a set together.

Perhaps adding "avx512_with_integer" as an alias for the "avx_skylake" would be worthwhile. (The hypothesis is this is the only AVX512 target that most people will use for shipping widely deployed software.) To take that further we could do:
    av512_float_only
    avx512_knl // If you're using this, you probably know what you're doing
    avx512_with_integer
    avx512_with_integer_cnl // Idea is Intel will maintain proper subsets from here :-)
though I am well aware I am tilting at a windmill.

**Comment by abadams** [comment]
I'm not entirely clear about the float/int distinction. It seems that
avx512f supports vectors of integers, but only 32 and 64-bit ones. avx512bw
adds support for 8 and 16-bit integers, and then avx512dq adds a richer set
of instructions for 32- and 64-bit integers. What a mess.

Anyway, I've pushed a naming change and some more comments.

**Comment by zvookin** [comment]
LGTM. (I'm tempted to suggest "AVX512_Base" for the current minimum AVX512 feature, but I'm not sure that is a win.)

File: test/correctness/simd_op_check.cpp [hunk 2]
```diff
@@ -1832,6 +1866,8 @@ int main(int argc, char **argv) {
         filter = argv[1];
     }
 
+    num_processes = 1;
```
Comment (dsharletg) [inline test/correctness/simd_op_check.cpp]: Did you intend to commit this change?

Comment (abadams) [inline test/correctness/simd_op_check.cpp]: I did not...

**Pull request merged**